kustomize/falco-driver/ebpf/daemonset.yaml

```diff
@@ -88,27 +84,38 @@ spec:
       - name: falco-driver-loader
         image: docker.io/falcosecurity/falco-driver-loader:master
         imagePullPolicy: IfNotPresent
-        args:
-          - ebpf
+        # args:
+        #   - ebpf
+        command: ["/bin/bash"]
+        args: # todo revert once new driver-loader is ready, symlinks didn't work for now
+          - -c
+          - >-
+            mkdir -p /usr/src/kernels/ &&
+            cp -r ${HOST_ROOT}/usr/src/kernels/$(uname -r) /usr/src/kernels/$(uname -r) &&
```

Comment:
@maxgio92 @LucaGuerra and @FedeDP it seems we have a bit of a regression. It used to work with the master `driver-loader` when we created this repo even when using `minikube`. Now there are still some hiccups with `HOST_ROOT`:

```
root@falco-driver-kmod-bt8t8:/# ls -l /lib/modules/6.2.15-100.fc36.x86_64/build 
lrwxrwxrwx. 1 root root 39 May 11  2023 /lib/modules/6.2.15-100.fc36.x86_64/build -> /usr/src/kernels/6.2.15-100.fc36.x86_64
```

But

```
/host/usr/src/kernels/6.2.15-100.fc36.x86_64
``

exists.

Suggestion: Can we not rely on `/lib/modules/` symlink and rather directly define the `KERNELDIR` for the make command?

Even better: Can we expose `KERNELDIR` as ENV variable to allow for better flexibility for possible issues with mounts or possible custom setups we can't yet anticipate?

Lastly, I would propose to alays test the driver-loader both with minikube and on actual Kubernetes nodes and not just with docker as there seems to be differences especially in terms of things that can go wrong. WDYT?

Reply:
Btw yes I was testing just with minikube but somehow symlinking didn't work .. strange, hence why I currently cp the kernel dir as workaround. Of course that has to go away asap.

Reply:
I don't get where it is breaking; you showed us just a `cp`.
Do you mean that then master driver-loader uses `HOST_ROOT` to access `/usr/src/kernels/...`?
Can you share driver loader logs?

Reply:
driver-loader apparently now looks in `/lib/modules/6.2.15-100.fc36.x86_64/build ` for the kernel sources and that symlinks to `/usr/src/kernels/6.2.15-100.fc36.x86_64` which does not exist since I mount it to `/host/usr/src/kernels/6.2.15-100.fc36.x86_64`.  It wasn't the case just 2.5 weeks ago before some recent changes in driver-loader.

I would prefer to avoid the defaults in the Makefile and consistently pass `KERNELDIR` to the make commands and we first check the dir is not empty or else give a clear warning why it's not working. WDYT? And as said a more robust solution with exposing some env variable or maybe can directly define `KERNELDIR` just so we have backup flexibility if all fails and the kernel sources are not mounted the right way but the end user can provide valid directories?

Reply:
> Can you share driver loader logs?

it just errored out from `/lib/modules/6.2.15-100.fc36.x86_64/build` nothing special about the error log really.

Reply:
> It wasn't the case just 2.5 weeks ago before some recent changes in driver-loader.

Just to help me understand this better, do you mean it used to work with the new falcoctl driver loader, or it never worked with it?

Reply:
> ```
> cp -r ${HOST_ROOT}/usr/src/kernels/$(uname -r) /usr/src/kernels/$(uname -r) &&
> ```
Testing on Ubuntu 22 LTS on AWS: when installing the kernel headers package, the sources are available at: 
* `/lib/modules/$(uname -r)/build` as symink to:
  * `/usr/src/linux-headers-$(uname -r)` (instead of `/usc/src/kernels/$(uname -r)`


At the same time, relying on `lib/modules/$(uname -r)` this way doesn't work as only `/usr/src/kernels` is mounted from the host fs.

Am I missing something?

Reply:
Rather than checking what changed or broke it, I would once more suggest to not rely on the default https://github.com/falcosecurity/libs/blob/master/driver/bpf/Makefile#L16 and instead add an option to pass the `KERNELDIR` to the make command similar to how I do it in the test VM setup here https://github.com/falcosecurity/libs/blob/99aa7d2161a65f03b4aab16382af92fe7e067e81/test/vm/scripts/compile_drivers.sh#L75-L76

I think we will then be stable and can hae a clean user interface requiring `${HOST_ROOT}/usr/src/kernels/` to contain the kernel sources needed to build kmod and ebpf.

Reply:
And @maxgio92 yes for ubuntu it's different https://github.com/falcosecurity/libs/blob/99aa7d2161a65f03b4aab16382af92fe7e067e81/test/vm/scripts/kernel_extract.sh hence an env variable would be best?

Reply:
I like the idea of exposing the `KERNELDIR` environment variable. My doubt is now: do we want to provide pre-filled values?
Because they could change as we saw between Linux distros

Reply:
I think that is a rather big (and impactful) change, too big to be made 2 weeks before a release.
Instead, i would love to understand whether (and when) we broke a working use case.
The old driver-loader script downloaded kernel sources and exported `KERNELDIR` during bpf probe compilation for minikube: https://github.com/falcosecurity/falco/blob/1b62b5ccd1c64cd972ef0252262075cbf42a130c/scripts/falco-driver-loader#L587
That is the exact same thing being done in current driver loader: 
* https://github.com/falcosecurity/falcoctl/blob/main/pkg/driver/distro/minikube.go#L100 (`env` is returned from `downloadKernelSrc`)
* `downloadKernelSrc` sets `KERNELDIR` env: https://github.com/falcosecurity/falcoctl/blob/main/pkg/driver/distro/distro.go#L353
* `bpf` builder sets env before running the build script: https://github.com/falcosecurity/falcoctl/blob/main/pkg/driver/type/bpf.go#L91

Reply:
Hi @FedeDP sharing some test VMs https://github.com/falcosecurity/cncf-green-review-testing/issues/7

Would you be able to test it with minikube? Out of curiosity have you ever tested the new driver-loader in Kubernetes? You could use the main branch or this branch after commenting back out the cmd override of the driver-loader initContainer.

CC @LucaGuerra

Reply:
@incertum briefly talking about that with @alacuku (that first introduced minikube prebuilt drivers support in old falco-driver-loader script), he told me that we never supported building drivers for minikube; that's why he added support for prebuilt artifacts.
I think this is the issue we are seeing.
If it worked a while ago, perhaps you were using a Falco version for which prebuilt artifacts were already present, so driver loader didn't attempt a build?

Reply:
I think that we can't rely on the `/usr/src/kernels` host directory as it depends on how the distro manages the kernel headers/sources directory (e.g. Ubuntu with headers package put them under _/usr/src/<package_name>_.

Moreover, the driver-loader [entrypoint](https://github.com/falcosecurity/falco/blob/master/docker/driver-loader/docker-entrypoint.sh) manages already the symlink creation under _/host/usr/src_ from _/usr/src_ (cc @FedeDP).
I propose this patch for the kmod loader manifest:

```diff
diff --git a/kustomize/falco-driver/kmod/daemonset.yaml b/kustomize/falco-driver/kmod/daemonset.yaml
index ff90798..9eb3f16 100644
--- a/kustomize/falco-driver/kmod/daemonset.yaml
+++ b/kustomize/falco-driver/kmod/daemonset.yaml
@@ -83,16 +83,8 @@ spec:
       - name: falco-driver-loader
         image: docker.io/falcosecurity/falco-driver-loader:master
         imagePullPolicy: IfNotPresent
-        # args: 
-        # - kmod
-        command: ["/bin/bash"]
-        args: # todo revert once new driver-loader is ready, symlinks didn't work for now
-          - -c
-          - >-
-            mkdir -p /usr/src/kernels/ &&
-            cp -r ${HOST_ROOT}/usr/src/kernels/$(uname -r) /usr/src/kernels/$(uname -r) &&
-            /usr/bin/falcoctl driver config --type kmod &&
-            /usr/bin/falcoctl driver install --compile=true --download=true
+        args:
+         - kmod
         securityContext:
           privileged: true
         volumeMounts:
@@ -109,15 +101,12 @@ spec:
           readOnly: true
         - mountPath: /host/lib/modules
           name: lib-modules
-        - mountPath: /host/usr/src/kernels/
-          name: usr-src-kernels-fs
+        - mountPath: /host/usr/src/
+          name: usr-src-fs
           readOnly: false # todo change after driver-loader fixes
         - mountPath: /host/etc
           name: etc-fs
           readOnly: true
-        env:
-        - name: HOST_ROOT
-          value: /host
       # todo @incertum upgrade to new tagged image after Falco 0.37.0 release
       - name: falcoctl-artifact-install
         image: docker.io/falcosecurity/falcoctl:0.6.2
@@ -148,9 +137,9 @@ spec:
       - name: lib-modules
         hostPath:
           path: /lib/modules
-      - name: usr-src-kernels-fs
+      - name: usr-src-fs
         hostPath:
-          path: /usr/src/kernels/
+          path: /usr/src/
       - name: etc-fs
         hostPath:
           path: /etc
```

Reply:
@maxgio92 Will implement those changes wrt to `/usr/src` as I frankly totally forgot about ubuntu distro as I live in RPM distro land ... I would still keep `HOST_ROOT` for technical clarity however if that's ok.

@FedeDP I used fedora where there is no pre-built drivers and it was building before the X-Mas break after I fixed the minikube start mounts, now same setup no success anymore, that was the puzzling part.

By the way I can't easily check on real Kubernetes with this exact setup, sorry about that. Would you be able to test on real Kubernetes (not minkube) and specifically test for building the driver on an RPM distro and ubuntu?

I will double-check again later with the VMs I linked to w/ minikube. There should be no reason why we can't build the driver also on minikube, especially if you confirm that the kernel sources were successfully mounted.

Reply:
Some updates:

@FedeDP First I am not so sure anymore before the XMas break I swear there was no pre-built driver for fedora ... anyhow more importantly I wanted to ask if y'all would accept the new bash command override for the driver-loader entrypoint?

That way I have 100% confirmed (on both ubuntu and fedora36 VMs https://github.com/falcosecurity/cncf-green-review-testing/issues/7) that we can build the driver from source even when using minikube. If we have these issues without minikube is a different discussion we can further double check in https://github.com/falcosecurity/falco/issues/2978.

In summary, I would very much like this setup to be compatible w/ minikube testing if acceptable.

@maxgio92 I added `/host/etc` to the falco container as that was another thing that was missing. Plus as stated above I tested on ubuntu now as well and I changed the `/host/usr` mount to the preferred setup.